convert/lib/nuke.py

```diff
@@ -0,0 +1,434 @@
+
+    def lambda_function(self, funcs: list[str]):
+        if not funcs:
+            return
+        existing_funcs = self._get_from_paged_list(
+            funcs,
+            self.awslambda.list_functions,
+            lambda x: [f["FunctionName"] for f in x["Functions"]],
+            "NextMarker",
+            next_token_arg="Marker",
+        )
+
+        if existing_funcs:
+            pprint({"Lambda Functions to delete": existing_funcs})
+
+            if self.delete:
+                for func in existing_funcs:
+                    self.awslambda.delete_function(FunctionName=func)
+
+    def lambda_layerversion(self, layerversions: list[str]):
+        if not layerversions:
+            return
+        layerversion_results = {}
+        for layerversion_arn in layerversions:
+            if re_result := re.match(
+                "arn:aws(?:-us-gov)?:lambda:[\w\-]+:\d+:layer:(\w*):(\d+)",
+                layerversion_arn,
+            ):
+                layer, version = re_result.groups(1)
+                version = int(version)
+                try:
+                    layer_result = self.awslambda.get_layer_version(
+                        LayerName=layer, VersionNumber=version
+                    )
+                    if layer_result["LayerVersionArn"] == layerversion_arn:
+                        layerversion_results[layerversion_arn] = {
+                            "LayerName": layer,
+                            "VersionNumber": version,
+                        }
+                except:
+                    pass
+
+        if layerversion_results:
+            pprint({"Lambda Layer Versions to delete": layerversion_results})
+
+            if self.delete:
+                for lv in layerversion_results.values():
+                    self.awslambda.delete_layer_version(**lv)
+
+    def ssm_parameter(self, parameters: list[str]):
+        if not parameters:
+            return
+        existing_parameters = self._get_from_paged_list(
+            parameters,
+            self.ssm.describe_parameters,
+            lambda x: [p["Name"] for p in x["Parameters"]],
+            "NextToken",
+        )
+
+        if existing_parameters:
+            pprint({"SSM Parameters to delete": existing_parameters})
+
+            if self.delete:
+                for p in existing_parameters:
+                    self.ssm.delete_parameter(Name=p)
+
+    def nuke(
+        self,
+        nuke_queue: dict[str, list[str]],
+        remove_security_group_references: bool = False,
+    ):
+        all_referenced_groups = {}
+        if security_groups := nuke_queue.get(cdk_ids.security_group.value):
+            for sg in security_groups:
+                if referenced_groups := [
+                    r
+                    for r in self.ec2.describe_security_group_rules()[
+                        "SecurityGroupRules"
+                    ]
+                    if r.get("ReferencedGroupInfo", {}).get("GroupId") == sg
```

Comment:
I think `describe_security_group_rules` takes a specific Group ID as a filter in the API call?

Reply:
This is seeing if the group id in the `ReferencedGroupInfo` part of the response is the same. We're specifically looking for groups rules for groups other than the one in question, that have a rule with a source of the group in question.